File: src/data/backup.ts
// src/data/backup.ts — 사용자 데이터 내보내기/가져오기
//
// Flutter 클라이언트로 데이터를 옮기는 **유일한 경로**다.
// WebView2 의 localStorage 를 직접 파헤치는 대신, 양쪽 앱이 읽고 쓰는
// 버전 있는 JSON 한 덩어리를 정본으로 삼는다.
//
// 진행 중인 대국(solo-snapshot)은 일부러 제외한다 — GameState 는 엔진 내부
// 표현이라 두 구현 사이에서 호환을 보장할 수 없고, 옮겨봐야 이어두기 어렵다.

export const BACKUP_FORMAT = 'svil-baduk-backup'
export const BACKUP_VERSION = 1

/** 백업에 담는 localStorage 키 — 진행 중 대국은 제외 */
export const BACKUP_KEYS = [
  'svil-baduk-settings',
  'svil-baduk-profile',
  'svil-baduk-learn-progress',
  'svil-baduk-solo-prefs',
] as const

export type BackupKey = (typeof BACKUP_KEYS)[number]

export interface BackupFile {
  format: typeof BACKUP_FORMAT
  version: number
  /** 내보낸 앱 버전 — 가져오는 쪽이 판단 근거로 쓴다 */
  appVersion: string
  savedAt: string
  data: Partial<Record<BackupKey, unknown>>
}

export type ImportResult =
  | { ok: true; restored: BackupKey[]; skipped: BackupKey[] }
  | { ok: false; reason: 'not_json' | 'not_backup' | 'unsupported_version' }

/** 현재 저장된 사용자 데이터를 한 덩어리로 모은다 */
export function buildBackup(appVersion: string): BackupFile {
  const data: Partial<Record<BackupKey, unknown>> = {}
  for (const key of BACKUP_KEYS) {
    const raw = localStorage.getItem(key)
    if (raw === null) continue
    try {
      data[key] = JSON.parse(raw)
    } catch {
      // 손상된 값은 조용히 건너뛴다 — 백업이 실패하는 것보다 낫다
    }
  }
  return {
    format: BACKUP_FORMAT,
    version: BACKUP_VERSION,
    appVersion,
    savedAt: new Date().toISOString(),
    data,
  }
}

/**
 * 백업을 되돌린다. 각 키는 원래 로더가 검증·마이그레이션하므로
 * 여기서는 형태만 확인하고 그대로 써 넣는다.
 */
export function restoreBackup(text: string): ImportResult {
  let parsed: unknown
  try {
    parsed = JSON.parse(text)
  } catch {
    return { ok: false, reason: 'not_json' }
  }
  if (!parsed || typeof parsed !== 'object') return { ok: false, reason: 'not_backup' }

  const file = parsed as Partial<BackupFile>
  if (file.format !== BACKUP_FORMAT) return { ok: false, reason: 'not_backup' }
  if (typeof file.version !== 'number' || file.version > BACKUP_VERSION) {
    return { ok: false, reason: 'unsupported_version' }
  }
  if (!file.data || typeof file.data !== 'object') return { ok: false, reason: 'not_backup' }

  const restored: BackupKey[] = []
  const skipped: BackupKey[] = []
  for (const key of BACKUP_KEYS) {
    const value = (file.data as Record<string, unknown>)[key]
    if (value === undefined) {
      skipped.push(key)
      continue
    }
    localStorage.setItem(key, JSON.stringify(value))
    restored.push(key)
  }
  return { ok: true, restored, skipped }
}

export function backupFilename(appVersion: string, now = new Date()): string {
  const stamp = now.toISOString().slice(0, 19).replace(/[:T]/g, '-')
  return `svil-baduk-backup-${appVersion}-${stamp}.json`
}

/** 브라우저에서 파일로 내려받기 */
export function downloadBackup(file: BackupFile): void {
  const blob = new Blob([JSON.stringify(file, null, 2)], { type: 'application/json' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = backupFilename(file.appVersion, new Date(file.savedAt))
  a.click()
  URL.revokeObjectURL(url)
}

File: app/lib/domain/backup.dart
// lib/domain/backup.dart — React 판 백업 파일 해석 (체크리스트 D5)
//
// React 웹(0.9.1+)의 "내 데이터 내보내기" 가 만든 JSON 을 읽는다.
// 형식 정본은 src/data/backup.ts — 버전 있는 한 덩어리가 두 구현 사이의
// 유일한 이전 경로다. 각 키의 내용 검증·마이그레이션은 원래 로더가 한다
// (settings 는 migrateSettings, profile 은 decodeProfile).

import 'dart:convert';

const String kBackupFormat = 'svil-baduk-backup';
const int kBackupVersion = 1;

/// 백업에 담기는 키 — React BACKUP_KEYS 와 동일해야 한다.
/// solo-prefs 는 Flutter 에 대응 저장소가 없어 무시된다.
const List<String> kBackupKeys = <String>[
  'svil-baduk-settings',
  'svil-baduk-profile',
  'svil-baduk-learn-progress',
];

sealed class BackupResult {
  const BackupResult();
}

/// 키 → 저장소에 그대로 쓸 JSON 문자열
final class BackupOk extends BackupResult {
  const BackupOk(this.values, {required this.appVersion});
  final Map<String, String> values;
  final String appVersion;
}

final class BackupFail extends BackupResult {
  const BackupFail(this.reasonKey);

  /// i18n 키: importFailedNotJson · importFailedNotBackup · importFailedVersion
  final String reasonKey;
}

BackupResult decodeBackup(String text) {
  final Object? parsed;
  try {
    parsed = jsonDecode(text);
  } on Object catch (_) {
    return const BackupFail('importFailedNotJson');
  }
  if (parsed is! Map) return const BackupFail('importFailedNotBackup');
  final Map<String, dynamic> j = parsed.cast<String, dynamic>();

  if (j['format'] != kBackupFormat) {
    return const BackupFail('importFailedNotBackup');
  }
  final Object? version = j['version'];
  if (version is! int || version > kBackupVersion) {
    return const BackupFail('importFailedVersion');
  }
  final Object? data = j['data'];
  if (data is! Map) return const BackupFail('importFailedNotBackup');

  final Map<String, String> values = <String, String>{};
  for (final String key in kBackupKeys) {
    final Object? v = data[key];
    if (v == null) continue;
    // localStorage 원문과 같은 형태로 되돌린다 — 로더가 그 형태를 기대한다
    values[key] = jsonEncode(v);
  }
  return BackupOk(values, appVersion: j['appVersion'] as String? ?? '');
}

/// Flutter 쪽 내보내기 — React restoreBackup 이 그대로 읽을 수 있는 형식
String encodeBackup({
  required String appVersion,
  required String savedAt,
  required Map<String, String> storedValues,
}) {
  final Map<String, Object?> data = <String, Object?>{};
  for (final MapEntry<String, String> e in storedValues.entries) {
    try {
      data[e.key] = jsonDecode(e.value);
    } on Object catch (_) {
      // 손상된 값은 조용히 건너뛴다 — 백업이 실패하는 것보다 낫다 (React 판과 동일)
    }
  }
  return const JsonEncoder.withIndent('  ').convert(<String, Object?>{
    'format': kBackupFormat,
    'version': kBackupVersion,
    'appVersion': appVersion,
    'savedAt': savedAt,
    'data': data,
  });
}
